project/Boilerplate.scala

```diff
@@ -0,0 +1,224 @@
+      }
+
+      def nullaryTuple(name: String) = {
+        val m = synTypes.map(tpe => s"${tpe}.${name}")
+        tuple(m)
+      }
+
+      block"""
+        |package algebra
+        |package std
+        |
+        |import algebra.ring.{EuclideanRing, Rig, Ring, Rng, Semiring}
+        |
+        |private[std] trait TupleInstances {
+        -  implicit def tuple${arity}Group[${`A..N`}](implicit ${constraints("Group")}): Group[${`(A..N)`}] =
+        -    new Group[${`(A..N)`}] {
+        -      def combine(x: ${`(A..N)`}, y: ${`(A..N)`}): ${`(A..N)`} = ${binTuple("combine")}
+        -      def empty: ${`(A..N)`} = ${nullaryTuple("empty")}
+        -      def inverse(x: ${`(A..N)`}): ${`(A..N)`} = ${unaryTuple("inverse")}
+        -    }
+        -
+        -  implicit def tuple${arity}Eq[${`A..N`}](implicit ${constraints("Eq")}): Eq[${`(A..N)`}] =
+        -    new Eq[${`(A..N)`}] {
+        -      def eqv(x: ${`(A..N)`}, y: ${`(A..N)`}): Boolean = ${binMethod("eqv").mkString(" && ")}
+        -    }
+        -
+        -  implicit def tuple${arity}EuclideanRing[${`A..N`}](implicit ${constraints("EuclideanRing")}): EuclideanRing[${`(A..N)`}] =
+        -    new EuclideanRing[${`(A..N)`}] {
+        -      def mod(x: ${`(A..N)`}, y: ${`(A..N)`}): ${`(A..N)`} = ${binTuple("mod")}
+        -      def negate(x: ${`(A..N)`}): ${`(A..N)`} = ${unaryTuple("negate")}
+        -      def one: ${`(A..N)`} = ${nullaryTuple("one")}
+        -      def plus(x: ${`(A..N)`}, y: ${`(A..N)`}): ${`(A..N)`} = ${binTuple("plus")}
+        -      def quot(x: ${`(A..N)`}, y: ${`(A..N)`}): ${`(A..N)`} = ${binTuple("quot")}
+        -      def times(x: ${`(A..N)`}, y: ${`(A..N)`}): ${`(A..N)`} = ${binTuple("times")}
+        -      def zero: ${`(A..N)`} = ${nullaryTuple("zero")}
+        -    }
+        -
+        -  implicit def tuple${arity}Field[${`A..N`}](implicit ${constraints("Field")}): Field[${`(A..N)`}] =
+        -    new Field[${`(A..N)`}] {
+        -      def div(x: ${`(A..N)`}, y: ${`(A..N)`}): ${`(A..N)`} = ${binTuple("div")}
+        -      def mod(x: ${`(A..N)`}, y: ${`(A..N)`}): ${`(A..N)`} = ${binTuple("mod")}
+        -      def negate(x: ${`(A..N)`}): ${`(A..N)`} = ${unaryTuple("negate")}
+        -      def one: ${`(A..N)`} = ${nullaryTuple("one")}
+        -      def plus(x: ${`(A..N)`}, y: ${`(A..N)`}): ${`(A..N)`} = ${binTuple("plus")}
+        -      def quot(x: ${`(A..N)`}, y: ${`(A..N)`}): ${`(A..N)`} = ${binTuple("quot")}
+        -      def times(x: ${`(A..N)`}, y: ${`(A..N)`}): ${`(A..N)`} = ${binTuple("times")}
+        -      def zero: ${`(A..N)`} = ${nullaryTuple("zero")}
+        -    }
+        -
+        -  implicit def tuple${arity}Monoid[${`A..N`}](implicit ${constraints("Monoid")}): Monoid[${`(A..N)`}] =
+        -    new Monoid[${`(A..N)`}] {
+        -      def combine(x: ${`(A..N)`}, y: ${`(A..N)`}): ${`(A..N)`} = ${binTuple("combine")}
+        -      def empty: ${`(A..N)`} = ${nullaryTuple("empty")}
+        -    }
```

Comment:
ideally we would compose `combineAll` in some way. At least add an issue. For many monoids the naive combineAll is very slow (for strings and lists it is `O(N^2)`, for Maps using an intermediate mutable map is much faster).